packages/effect/src/internal/trie/node.ts

```diff
@@ -0,0 +1,11 @@
+/** @internal */
+export class Node<out V> {
+  constructor(
+    public key: string,
+    public count: number,
+    public value: V | undefined = undefined,
+    public left: Node<V> | undefined = undefined,
+    public mid: Node<V> | undefined = undefined,
+    public right: Node<V> | undefined = undefined
+  ) {}
+}
```

Comment:
Not sure how much this needs to be optimized for performance, but if this is just a value bag, you don't need a class for it. Especially not if it's only used internally. A plain object & interface should give you significantly better performance (easy to proof that with a benchmark).

Considering the simplicity of this, you might just want to create these plain objects inline even (a `make` function wouldn't really add any value imho).

Reply:
And when I say "significant", this is what I mean :-)

https://jsbenchmark.com/#eyJjYXNlcyI6W3siaWQiOiJweXREb012SlFaQXNvbGtrdVlka2MiLCJjb2RlIjoiZnVuY3Rpb24gUG9pbnQoeCwgeSkge1xuICByZXR1cm4geyB4LCB5IH1cbn1cblxuZm9yIChsZXQgaSA9IDA7IGkgPD0gMTAwMDA7IGkrKykge1xuICBQb2ludCgxMjMsIDQ1Nilcbn0iLCJuYW1lIjoiT2JqZWN0IExpdGVyYWwifSx7ImlkIjoiS21SMjdLeWxlR2pQMHlDb0lrMDQ5IiwiY29kZSI6ImNsYXNzIFBvaW50IHtcbiAgY29uc3RydWN0b3IoeCwgeSkge1xuICAgIHRoaXMueCA9IHhcbiAgICB0aGlzLnkgPSB5XG4gIH1cbn1cblxuZm9yIChsZXQgaSA9IDA7IGkgPD0gMTAwMDA7IGkrKykge1xuICBuZXcgUG9pbnQoMTIzLCA0NTYpXG59IiwibmFtZSI6IkNsYXNzIn1dLCJjb25maWciOnsibmFtZSI6Ik9iamVjdCBMaXRlcmFsIHZzLiBDbGFzcyIsInBhcmFsbGVsIjp0cnVlLCJnbG9iYWxUZXN0Q29uZmlnIjp7ImRlcGVuZGVuY2llcyI6W119LCJkYXRhQ29kZSI6IiJ9fQ==

Reply:
![object-literal-vs -class-(10 000-iterations)](https://github.com/Effect-TS/effect/assets/1172528/60ea97d7-fb2b-4c2c-94e3-3a267feeda00)